server/src/main/java/org/elasticsearch/lucene/util/automaton/MinimizationOperations.java

```diff
@@ -0,0 +1,329 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.lucene.util.automaton;
+
+import org.apache.lucene.internal.hppc.IntArrayList;
+import org.apache.lucene.internal.hppc.IntCursor;
+import org.apache.lucene.internal.hppc.IntHashSet;
+import org.apache.lucene.util.automaton.Automaton;
+import org.apache.lucene.util.automaton.Operations;
+import org.apache.lucene.util.automaton.Transition;
+
+import java.util.BitSet;
+import java.util.LinkedList;
+
+/**
+ * Operations for minimizing automata.
+ * <p>
+ * Copied of Lucene's AutomatonTestUtil
```

Comment:
Lets add a bit of background here on why we brought this back? 
I.e. "Lucene removed the minimization logic but Elasticsearch still requires it. Elasticsearch makes heavy use of automata for checking permissions in the security codebase which consumes a prohibitive amount of heap for automata without minimization."

Reply:
Oh ya, good call. Pushed https://github.com/elastic/elasticsearch/pull/119309/commits/eb8ef47f6e53f3551fe954fa09ddb774ef4bbe2a